Platform/OpenNtfsDxe/Data.c

```diff
@@ -0,0 +1,566 @@
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  if ((Offset + Length) > NonRes->RealSize) {
+    DEBUG((DEBUG_INFO, "NTFS: Read out of range\n"));
+    FreePool (Runlist);
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  if (((NonRes->Flags & FLAG_COMPRESSED) != 0) && // ((Attr->Flags & NTFS_AF_GPOS) == 0))
+      (NonRes->Type == AT_DATA)) {
+    Status = Decompress (Runlist, Offset, Length, Dest);
+
+    FreePool (Runlist);
+    return Status;
+  }
+
+  while (Runlist->NextVcn <= Runlist->TargetVcn) {
+    Status = ReadRunListElement (Runlist);
+    if (EFI_ERROR(Status)) {
+      FreePool (Runlist);
+      return EFI_VOLUME_CORRUPTED;
+    }
+  }
+
+  // if (Attr->Flags & NTFS_AF_GPOS) {
+  //   OffsetInsideCluster = Offset & (mClusterSize - 1);
+  //
+  //   Sector0 = ((Runlist->TargetVcn - Runlist->CurrentVcn + Runlist->CurrentLcn)
+  //             * mClusterSize + OffsetInsideCluster) / mSectorSize;
+  //
+  //   Sector1 = Sector0 + 1;
+  //   if (Sector1 == ((Runlist->NextVcn - Runlist->CurrentVcn + Runlist->CurrentLcn)
+  //                  * mClusterSize) / mSectorSize) {
+  //     Status = ReadRunListElement (Runlist);
+  //     if (EFI_ERROR(Status)) {
+  //       FreePool (Runlist);
+  //       return EFI_VOLUME_CORRUPTED;
+  //     }
+  //
+  //     Sector1 = Runlist->CurrentLcn * mClusterSize / mSectorSize;
+  //   }
+  //
+  //   *(UINT32 *) Dest = (UINT32) Sector0;
+  //   *(UINT32 *) (Dest + 4) = (UINT32) Sector1;
+  //
+  //   FreePool (Runlist);
+  //   return EFI_SUCCESS;
+  // }
+
+  Status = ReadClusters (
+    Runlist,
+    Offset,
+    Length,
+    Dest
+    );
+
+  FreePool (Runlist);
+  return Status;
+}
+
+STATIC
+UINT64
+ReadField (
+  IN CONST UINT8 *Run,
+  IN INT32       FieldSize,
+  IN BOOLEAN     Signed
+  )
+{
+  UINT64 Value = 0;
+  //
+  // Offset to the starting LCN of the previous element is a signed value.
+  // So we must check the most significant bit.
+  //
+  if (Signed && FieldSize && (Run[FieldSize - 1] & 0x80)) {
+    Value = (UINT64) (-1);
+  }
+
+  CopyMem (&Value, Run, FieldSize);
+
+  return Value;
+}
+
+/**
+   Table 4.10. Layout of a Data Run
+   ____________________________________________________________________
+   Offset | Size | Description
+   --------------------------------------------------------------------
+   0      | 0.5   | F=Size of the Offset field
+   0.5    | 0.5   | L=Size of the Length field
+   1      | L     | Length of the run
+   1+L    | F     | Offset to the starting LCN of the previous element (signed)
+   --------------------------------------------------------------------
+**/
+EFI_STATUS
+EFIAPI
+ReadRunListElement (
+  IN OUT RUNLIST  *Runlist
+  )
+{
+  UINT8              LengthSize;
+  UINT8              OffsetSize;
+  UINT64             OffsetLcn;
+  UINT8              *Run;
+  ATTR_HEADER_NONRES *Attr;
+  UINT64             BufferSize;
+
+  Run = Runlist->NextDataRun;
+  BufferSize = mFileRecordSize - (Run - Runlist->Attr->BaseMftRecord->FileRecord);
+
+  retry:
```

Comment:
Try not to label.

Reply:
This will require substantial refactoring. https://github.com/acidanthera/OpenCorePkg/pull/332#discussion_r844984887